finetuner/experiment.py

```diff
@@ -117,6 +120,75 @@ def delete_runs(self):
         """Delete every run inside the experiment."""
         self._client.delete_runs(experiment_name=self._name)
 
+    def _prepare_data(
+        self,
+        train_data: Union[DocumentArray, str],
+        run_name: str,
+        eval_data: Optional[Union[DocumentArray, str]] = None,
+        eval_callback: Optional[EvaluationCallback] = None,
+    ):
+        """Upload data to Hubble and returns their names.
+
+        Uploads all data needed for fine-tuning - training data,
+        evaluation data and query/index data for `EvaluationCallback`.
+
+        Checks not to upload same dataset twice.
+
+        :param train_data: Either a `DocumentArray` for training data or a
+            name of the `DocumentArray` that is pushed on Hubble.
+        :param run_name: Name of the run.
+        :param train_data: Either a `DocumentArray` for evaluation data or a
+            name of the `DocumentArray` that is pushed on Hubble.
+        :param eval_callback: Evaluation callback that contains query and index data
+            and gets modified in-place.
+        :return: Name(s) of the uploaded data.
+        """
+        train_data_name = push_data_to_hubble(
+            client=self._client,
+            data=train_data,
+            data_type=TRAIN_DATA,
+            experiment_name=self._name,
+            run_name=run_name,
+        )
+        eval_data_name = None
+        if eval_data:
+            eval_data_name = push_data_to_hubble(
+                client=self._client,
+                data=eval_data,
+                data_type=EVAL_DATA,
+                experiment_name=self._name,
+                run_name=run_name,
+            )
+        if eval_callback:
+            if eval_callback.query_data == train_data:
+                eval_callback.query_data = train_data_name
+            elif eval_callback.query_data == eval_data:
+                eval_callback.query_data = eval_data_name
+            else:
+                eval_callback.query_data = push_data_to_hubble(
+                    client=self._client,
+                    data=eval_callback.query_data,
+                    data_type=QUERY_DATA,
+                    experiment_name=self._name,
+                    run_name=run_name,
+                )
+
+            if not eval_callback.index_data:
+                pass
+            elif eval_callback.index_data == train_data:
+                eval_callback.index_data = train_data_name
+            elif eval_callback.index_data == eval_data:
+                eval_callback.index_data = eval_data_name
+            else:
+                eval_callback.index_data = push_data_to_hubble(
+                    client=self._client,
+                    data=eval_callback.index_data,
+                    data_type=INDEX_DATA,
+                    experiment_name=self._name,
+                    run_name=run_name,
+                )
+        return train_data_name, eval_data_name
+
```

Comment:
The only thing that should be done here is to check whether eval, query and index data are given and if they are docarrays or strings. If they are docarrays they should be pushed.
```suggestion
        eval_data_name = None
        query_data_name = None
        index_data_name = None
        if eval_data:
            eval_data_name = push_data_to_hubble(
                client=self._client,
                data=eval_data,
                data_type=EVAL_DATA,
                experiment_name=self._name,
                run_name=run_name,
            )
         if query_data:
            query_data_name = push_data_to_hubble(
                client=self._client,
                data=query_data,
                data_type=QUERY_DATA,
                experiment_name=self._name,
                run_name=run_name,
            )
         if index_data:
            index_data_name = push_data_to_hubble(
                client=self._client,
                data=index_data,
                data_type=INDEX_DATA,
                experiment_name=self._name,
                run_name=run_name,
            )
        
        return train_data_name, eval_data_name, query_data_name, index_data_name

```

Reply:
and we don't care that the same data could possibly be uploaded several times?